crates/dbsp/src/trace/spine_async/list_merger.rs

```diff
@@ -102,100 +131,309 @@ where
+        } else {
+            builder.push_val(self.cursors[indexes.first().unwrap()].val());
         }
         *fuel -= 1;
-        self.cursor.step_val();
-        any
+        true
     }
+}
 
-    /// Extracts merged results.
-    ///
-    /// This method should only be called after `work` has been called and
-    /// has not brought `fuel` to zero. Otherwise, the merge is still in
-    /// progress.
-    pub fn done(self) -> B {
-        let lower = self.batches[1..]
-            .iter()
-            .fold(self.batches[0].lower().to_owned(), |lower, batch| {
-                batch.lower().meet(lower.as_ref())
-            });
-        let upper = self.batches[1..]
-            .iter()
-            .fold(self.batches[0].upper().to_owned(), |upper, batch| {
-                batch.upper().join(upper.as_ref())
-            });
+fn find_min_indexes<Item>(mut iterator: impl Iterator<Item = (usize, Item)>) -> IndexSet
+where
+    Item: Ord,
+{
+    let (min_index, mut min_value) = iterator.next().unwrap();
+    let mut min_indexes = IndexSet::for_index(min_index);
+
+    for (index, value) in iterator {
+        match value.cmp(&min_value) {
+            Ordering::Less => {
+                min_value = value;
+                min_indexes = IndexSet::for_index(index);
+            }
+            Ordering::Equal => {
+                min_indexes.add(index);
+            }
+            Ordering::Greater => (),
+        }
+    }
+    min_indexes
+}
+
+/// A set of indexes (numbers) in the range 0..=63.
+#[derive(Copy, Clone, PartialEq, Eq)]
+struct IndexSet(u64);
+
+impl IndexSet {
+    /// Returns a bit-set with only `index` set to 1.
+    fn for_index(index: usize) -> Self {
+        debug_assert!(index < 64);
+        Self(1 << index)
+    }
+
+    /// Returns an empty set.
+    fn empty() -> Self {
+        Self(0)
+    }
+
+    /// Returns true if the set is empty.
+    fn is_empty(&self) -> bool {
+        self.0 == 0
+    }
+
+    /// Returns true if the set contains exactly one index.
+    #[allow(dead_code)]
+    fn is_singleton(&self) -> bool {
+        self.0.is_power_of_two()
+    }
+
+    /// Returns true if the set contains exactly zero or one indexes.
+    fn is_short(&self) -> bool {
+        self.without_first().is_empty()
+    }
+
+    /// Returns true if the set contains two or more indexes.
+    fn is_long(&self) -> bool {
+        !self.is_short()
+    }
+
+    /// Returns the number of indexes in the set.
+    fn len(&self) -> usize {
+        self.0.count_ones() as usize
+    }
+
+    /// Returns the smallest index in the set.
+    fn first(&self) -> Option<usize> {
+        if self.0 == 0 {
+            None
+        } else {
+            Some(self.0.trailing_zeros() as usize)
+        }
+    }
+
+    // Returns the set with the first index removed. If this set is empty,
+    // returns an empty set.
+    fn without_first(&self) -> Self {
+        Self(self.0 & (self.0.wrapping_sub(1)))
```

Comment:
this is very clever

Reply:
I originally got it from Hacker's Delight (https://en.wikipedia.org/wiki/Hacker%27s_Delight), a book that I heartily recommend.